openedx/core/lib/call_stack_manager.py

```diff
@@ -0,0 +1,138 @@
+"""
+Get call stacks of Model Class
+in three cases-
+1. QuerySet API
+2. save()
+3. delete()
+
+classes:
+CallStackManager -  stores all stacks in global dictionary and logs
+CallStackMixin - used for Model save(), and delete() method
+
+Functions:
+capture_call_stack - global function used to store call stack
+
+Decorators:
+donottrack - mainly for the places where we know the calls. This decorator will let us not to track in specified cases
+How to use-
+1. Import following in the file where class to be tracked resides
+    from openedx.core.lib.call_stack_manager import CallStackManager, CallStackMixin
+2. Override objects of default manager by writing following in any model class which you want to track-
+    objects = CallStackManager()
+3. For tracking Save and Delete events-
+    Use mixin called "CallStackMixin"
+    For ex.
+        class StudentModule(CallStackMixin, models.Model):
+
+Note -
+1.Format for STACK_BOOK
+{
+    "modelclass1":
+        [[(frame 1),(frame 2)],
+         [(frame 11),(frame21)]]
+    "modelclass2":
+        [[(frame 3),(frame 4)],
+         [(frame 6),(frame 5)]]
+
+}
+where frame is a tuple of
+(file path, Line Number, Context)
+"""
+
+import logging
+import traceback
+import re
+import collections
+from django.db.models import Manager
+
+log = logging.getLogger(__name__)
+
+# Module Level variables
+# dictionary which stores call stacks.
+# { "ModelClasses" : [ListOfFrames]}
+# Frames - ('FilePath','LineNumber','Context')
+# ex. {"<class 'courseware.models.StudentModule'>" : [[(file,line number,context),(---,---,---)],
+#                                                     [(file,line number,context),(---,---,---)]]}
+STACK_BOOK = collections.defaultdict(list)
+
+# filter to trickle down call stacks
+EXCLUDE = ['^.*python2.7.*$', '^.*call_stack_manager.*$', '^.*<exec_function>.*$', '^.*exec_code_object.*$']
+REGULAR_EXPS = [re.compile(x) for x in EXCLUDE]
+
+TRACK_FLAG = True
+
+
+def capture_call_stack(current_model):
+    """
+    stores customised call stacks in global dictionary `STACK_BOOK`, and logs it.
+
+    Arguments:
+    current_model - Name of the model class
+    """
+
+    # holds temporary callstack
+    # frame[0][6:-1] -> File name along with path
+    # frame[1][6:] -> Line Number
+    # frame[2][3:] -> Context
+    temp_call_stack = [(frame[0][6:-1],
+                        frame[1][6:],
+                        frame[2][3:])
+                       for frame in [stack.replace("\n", "").strip().split(',') for stack in traceback.format_stack()]
+                       if not any(reg.match(frame[0]) for reg in REGULAR_EXPS)]
+
+    # avoid duplication.
+    if temp_call_stack not in STACK_BOOK[current_model]:
+        STACK_BOOK[current_model].append(temp_call_stack)
+        log.info("logging new call in global stack book, for %s", current_model)
+        log.info(STACK_BOOK)
```

Comment:
I can do that too. However, `STACK_BOOK` is a module level dictionary which keeps track of all "to-be-tracked" model classes, and their respective call stacks. If we only log newly-added stack trace, we wont be able to get whole `STACK_BOOK`. It would just maintain the `STACK_BOOK` without any use. In such case, how do we capture the whole `STACK_BOOK`?